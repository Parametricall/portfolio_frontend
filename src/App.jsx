import React, { useEffect, useState } from "react";
import { BrowserRouter as Router, Switch, Route } from "react-router-dom";
import { Link as RouterLink } from "react-router-dom";
import { LinkContainer } from "react-router-bootstrap";
import LandingPage from "./components/LandingPage";
import User from "./user/components";
import Navbar from "react-bootstrap/Navbar";
import Nav from "react-bootstrap/Nav";
import Login from "./user/components/Login";
import Logout from "./user/components/Logout";
import { postData } from "./utilities";
import { GET_TOKEN_URL } from "./constants";
import { NavDropdown } from "react-bootstrap";
import { Redirect } from "react-router-dom";
import Cookbook from "./cookbook/components";
import ProtectedRoute from "./components/ProtectedRoute";
import SnakeGame from "./snakeGame/components/SnakeGame";
import { AppBar, IconButton, Link, makeStyles, Menu, MenuItem, Toolbar, Typography } from "@material-ui/core";
import { AccountCircle } from "@material-ui/icons";


const useStyles = makeStyles((theme) => ({
  title: {
    display: "none",
    [theme.breakpoints.up("sm")]: {
      display: "block"
    }
  },
  grow: {
    flexGrow: 1
  },
  linkHover: {
    "&:hover": {
      color: "unset",
      textDecorationColor: "unset"
    }
  }
}));


function App() {
  const classes = useStyles();
  const [user, setUser] = useState(null);
  const [anchorEl, setAnchorEl] = React.useState(null);
  const isMenuOpen = Boolean(anchorEl);

  useEffect(() => {
    const loggedInUser = localStorage.getItem("user");
    if (loggedInUser) {
      const foundUser = JSON.parse(loggedInUser);
      setUser(foundUser);
    }
  }, []);

  const logout = () => {
    localStorage.clear();
    setUser(null);
  };

  const handleLogin = async (e, username, password) => {
    e.preventDefault();
    const user = { username, password };
    localStorage.clear();
    const response = await postData(GET_TOKEN_URL, user).then((json) => json);
    setUser(response);
    localStorage.setItem("user", JSON.stringify(response));
  };

  const handleProfileMenuOpen = (event) => {
    setAnchorEl(event.currentTarget);
  };

  const handleMenuClose = () => {
    setAnchorEl(null);
  };

  const menuId = "primary-search-account-menu";
  const renderMenu = (
    <Menu
      anchorEl={anchorEl}
      anchorOrigin={{ vertical: "top", horizontal: "right" }}
      id={menuId}
      keepMounted
      transformOrigin={{ vertical: "top", horizontal: "right" }}
      open={isMenuOpen}
      onClose={handleMenuClose}
    >
      <MenuItem component={RouterLink} to={`/users/${user?.user_id}`} onClick={handleMenuClose}>Profile</MenuItem>
      <MenuItem component={RouterLink} to="/users" onClick={handleMenuClose}>Users</MenuItem>
      <MenuItem component={RouterLink} to="/logout" onClick={handleMenuClose}>Logout</MenuItem>
    </Menu>
  );

  return (
    <Router>
      <div className={classes.grow}>
        <AppBar position="static">
          <Toolbar>
            <Typography className={classes.title} variant="h6" noWrap>
              <Link className={classes.linkHover} color="inherit" component={RouterLink} to="/home">
                Home
              </Link>
            </Typography>
            <div className={classes.grow} />
            <div>
              <IconButton
                edge="end"
                aria-label="account of current user"
                aria-controls={menuId}
                aria-haspopup="true"
                onClick={handleProfileMenuOpen}
                color="inherit"
              >
                <AccountCircle />
              </IconButton>
            </div>
          </Toolbar>
        </AppBar>
        {renderMenu}
      </div>

      <Switch>
        <Route exact path="/home">
          <LandingPage />
        </Route>
        <Route path="/cookbook">
          <Cookbook />
        </Route>
        <Route path="/snake_game">
          <SnakeGame />
        </Route>
        <ProtectedRoute path="/users">
          <User />
        </ProtectedRoute>
        <Route path="/login">
          <Login handleLogin={handleLogin} setUser={setUser} />
        </Route>
        <Route path="/logout">
          <Logout logout={logout} />
        </Route>
        <Route path="/">
          <Redirect to="/home" />
        </Route>
      </Switch>
    </Router>
  );
}

export default App;
