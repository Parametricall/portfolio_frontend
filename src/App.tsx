import React, { useEffect } from 'react';
import {
    BrowserRouter as Router, Link as RouterLink, Redirect, Route, Switch,
} from 'react-router-dom';
import {
    AppBar, IconButton, Link, makeStyles, Menu, MenuItem, Toolbar, Typography,
} from '@material-ui/core';
import { AccountCircle } from '@material-ui/icons';
import { useDispatch, useSelector } from 'react-redux';
import LandingPage from './components/LandingPage';
import User from './user/components';
import Login from './user/components/Login';
import Logout from './user/components/Logout';
import { fetchJsonData, postData } from './utilities';
import { GET_TOKEN_URL, RETRIEVE_USERS_URL } from './constants';
import Cookbook from './cookbook/components';
import SnakeGame from './snakeGame/components/SnakeGame';
import { setUser2 } from './reduxStore/actions';
import UserPermissions from './components/UserPermissions';

const useStyles = makeStyles((theme) => ({
    title: {
        display: 'none',
        [theme.breakpoints.up('sm')]: {
            display: 'block',
        },
    },
    grow: {
        flexGrow: 1,
    },
    linkHover: {
        '&:hover': {
            color: 'unset',
            textDecorationColor: 'unset',
        },
    },
}));

async function updateStoreFromUserToken(dispatch) {
    const token = localStorage.getItem('user');
    // const user = sessionStorage.getItem('user');

    if (token) {
        const userId = JSON.parse(token).user_id;
        const userObj = await fetchJsonData(
            `${RETRIEVE_USERS_URL}${userId}/`,
            'GET',
            null,
        );
        // todo: handle error response here
        dispatch(setUser2({ ...userObj, isAuthenticated: true }));
    }
}

interface UserInterface {
    user_permissions: string[],
    isAuthenticated: boolean
    id: number,
}

function App() {
    const classes = useStyles();
    // const [user, setUser] = useState(null);
    const [anchorEl, setAnchorEl] = React.useState(null);
    const isMenuOpen = Boolean(anchorEl);

    const currentUser = useSelector((state:{user: UserInterface}) => state.user);
    const dispatch = useDispatch();

    useEffect(() => {
        if (!currentUser.isAuthenticated) {
            updateStoreFromUserToken(dispatch).then();
        }
    });

    // useEffect(() => {
    //     const getUser = async () => {
    //
    //         if (!currentUser.groups) {
    //             const action = await getuserDetails();
    //             if (action.action === "login") {
    //                 return <Redirect to="/login" />;
    //             } else {
    //                 dispatch(setUser2(action.user));
    //             }
    //         }
    //     };
    //     getUser();
    // });

    // useEffect(() => {
    //     const loggedInUser = localStorage.getItem("user");
    //     if (loggedInUser) {
    //         const foundUser = JSON.parse(loggedInUser);
    //         setUser(foundUser);
    //     }
    // }, []);

    // const logout = () => {
    //     localStorage.clear();
    //     dispatch(setUser2({ isAuthenticated: false }));
    //     // setUser(null);
    // };

    const handleLogin = async (e, username, password) => {
        e.preventDefault();
        const userCred = { username, password };
        localStorage.clear();
        const response = await postData(GET_TOKEN_URL, userCred).then((json) => json);
        // setUser(response);
        localStorage.setItem('user', JSON.stringify(response));
    };

    const handleProfileMenuOpen = (event) => {
        setAnchorEl(event.currentTarget);
    };

    const handleMenuClose = () => {
        setAnchorEl(null);
    };

    const menuId = 'primary-search-account-menu';
    const renderMenu = (
        <Menu
            anchorEl={anchorEl}
            anchorOrigin={{ vertical: 'top', horizontal: 'right' }}
            id={menuId}
            keepMounted
            transformOrigin={{ vertical: 'top', horizontal: 'right' }}
            open={isMenuOpen}
            onClose={handleMenuClose}
        >
            <UserPermissions showIfAuthenticated={false}>
                <MenuItem component={RouterLink} to="/login" onClick={handleMenuClose}>
                    Login
                </MenuItem>
            </UserPermissions>
            <UserPermissions permissions={['auth.view_user']}>
                <MenuItem
                    component={RouterLink}
                    to={`/users/${currentUser?.id}`}
                    onClick={handleMenuClose}
                >
                    Profile
                </MenuItem>
            </UserPermissions>
            <UserPermissions permissions={['auth.add_user']}>
                <MenuItem component={RouterLink} to="/users" onClick={handleMenuClose}>
                    Users
                </MenuItem>
            </UserPermissions>
            <UserPermissions>
                <MenuItem component={RouterLink} to="/logout" onClick={handleMenuClose}>
                    Logout
                </MenuItem>
            </UserPermissions>
        </Menu>
    );

    return (
        <Router>
            <div className={classes.grow}>
                <AppBar position="static">
                    <Toolbar>
                        <Typography className={classes.title} variant="h6" noWrap>
                            <Link
                                className={classes.linkHover}
                                color="inherit"
                                component={RouterLink}
                                to="/home"
                            >
                                Home
                            </Link>
                        </Typography>
                        <div className={classes.grow} />
                        <div>
                            <IconButton
                                edge="end"
                                aria-label="account of current user"
                                aria-controls={menuId}
                                aria-haspopup="true"
                                onClick={handleProfileMenuOpen}
                                color="inherit"
                            >
                                <AccountCircle />
                            </IconButton>
                        </div>
                    </Toolbar>
                </AppBar>
                {renderMenu}
            </div>

            <Switch>
                <Route exact path="/home">
                    <LandingPage />
                </Route>
                <Route path="/cookbook">
                    <Cookbook />
                </Route>
                <Route path="/snake_game">
                    <SnakeGame />
                </Route>
                <Route path="/users">
                    <User />
                </Route>
                <Route path="/login">
                    <Login
                        handleLogin={handleLogin}
                        setUser={() => {
                        }}
                    />
                </Route>
                <Route path="/logout">
                    <Logout />
                </Route>
                <Route path="/">
                    <Redirect to="/home" />
                </Route>
            </Switch>
        </Router>
    );
}

export default App;
