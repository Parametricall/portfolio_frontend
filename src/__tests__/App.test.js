import React from 'react';
import { shallow } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';

import App from '../App';


describe('App', () => {
  const wrapper = shallow(<MemoryRouter><App/></MemoryRouter>);
  const app = wrapper.find(App).dive();

  it('renders correctly', () => {
    expect(app).toMatchSnapshot();
  });

  describe('when no user is logged in', () => {
    it('displays "Login" in the nav bar', () => {
      expect(app.find('.app-login').exists()).toBe(true)
    })
  })
});