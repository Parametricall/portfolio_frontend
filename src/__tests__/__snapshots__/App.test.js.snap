// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `
<BrowserRouter>
  <Navbar
    bg="light"
    collapseOnSelect={false}
    expand="lg"
    variant="light"
  >
    <NavbarToggle
      aria-controls="basic-navbar-nav"
      label="Toggle navigation"
    />
    <NavbarCollapse
      id="basic-navbar-nav"
    />
    <Nav
      className="mr-auto"
      fill={false}
      justify={false}
    >
      <withRouter(LinkContainer)
        to="/home"
      >
        <NavLink
          as={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "displayName": "SafeAnchor",
              "render": [Function],
            }
          }
          disabled={false}
        >
          Home
        </NavLink>
      </withRouter(LinkContainer)>
      <withRouter(LinkContainer)
        className="app-login"
        to="/login"
      >
        <NavLink
          as={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "displayName": "SafeAnchor",
              "render": [Function],
            }
          }
          disabled={false}
        >
          Login
        </NavLink>
      </withRouter(LinkContainer)>
    </Nav>
  </Navbar>
  <Switch>
    <Route
      exact={true}
      path="/home"
    >
      <LandingPage />
    </Route>
    <Route
      path="/cookbook"
    >
      <User />
    </Route>
    <Route
      path="/users"
    >
      <User />
    </Route>
    <Route
      path="/login"
    >
      <Login
        handleLogin={[Function]}
      />
    </Route>
    <Route
      path="/logout"
    >
      <Logout
        logout={[Function]}
      />
    </Route>
    <Route
      path="/"
    >
      <Redirect
        to="/home"
      />
    </Route>
  </Switch>
</BrowserRouter>
`;
